### DISCUSSION TITLE
Convert PNG to YUVIL 4:2:0 8-bit

### DISCUSSION THREAD

**Discussion by Eutechnyx** [ImageMagick/ImageMagick#6276] [Help] [answered]
### ImageMagick version

7.1.1-Q16

### Operating system, version and so on

Windows 10

### Description

Need help converting PNG file to a YUV(IL) 4:2:0 8-bit RAW.
I have an example of source RAW of a frame from device. Decoding it using parameters:

`magick -size 352x80 -depth 8 yuv:raw.yuv png:image.png`

gave me a correct file.
But while encoding it back using:

`magick convert image.png -depth 8 -colorspace YUV -sampling-factor 2x2 -size 352x80 -interlace plane -define interlace:plane=interleave image.yuv`

i'm getting standart YUV 4:2:0 file, not interleaved, like using regular convertion option:

`magick convert image.png -depth 8 -colorspace YUV -sampling-factor 2x2 -size 352x80 image.yuv`

If grayscale image is converted - everything is displayed fine, but i need colors.

When displaying on a device, it looks like this:
[Image](https://drive.google.com/file/d/1vuqE4DJl8sVdFBJypOsZK8GWMCu1xzky/view?usp=share_link)
But should be displayed this way:
[Image](https://drive.google.com/file/d/1AFAyhBhcKuzABsortQDRcSDVXqyQpCTM/view?usp=share_link)

I'm checking all those RAW images after convertion using [YUView](https://github.com/IENT/YUView) with these options: 352x80; YUV 4:2:0; 8-Bit depth; Chroma Offset X - 0; Y - 1/2; Planar; YUV plane order: YUV; UV(A) Interleaved. These settings match the actual displaying on my device.
[Settings](https://drive.google.com/file/d/1F6NqegMZ3CfJWYfs_yoW1wQcLofYb-P7/view?usp=share_link)

What am i doing wrong and could it be done using ImageMagick at all?

**Comment by snibgo** [comment]
I don't know what "`UV(A) Interleaved`" means, but it might correspond to the IM format `UYVY`, which is "Interleaved YUV raw image". Note that in IM, "interleaved" is not the same as "interlace".

As far as I can see, your "`-define interlace:plane=interleave`" has no effect for any format.

**Comment by Eutechnyx** [comment]
I'm guessing that means it should be Y planar and UV are interleaved.
UYVY is 4:2:2 - tried, not my variant.

**Comment by snibgo** [comment]
I suggest you link to your `raw.yuv`. You can zip it and post it here. Then we can convert it to PNG, and try various options to attempt the round trip.

**Comment by Eutechnyx** [comment]
Okay, just captured a new one after turning it on - looks like a booting process logo.
[YUV](https://drive.google.com/file/d/1SKIkFtOAP3xdRHHQaCLWHeoYwNxytczD/view?usp=share_link)

**Comment by snibgo** [comment]
As you should realise from your first post, to convert that yuv file to PNG we need to know the width and height in pixels, and depth in bits.

**Comment by Eutechnyx** [comment]
I thought i wrote everything related to this in my first post. Size 352x80, Depth 8.

**Comment by snibgo** [comment]
The image was different, so I wasn't sure if the size and depth were the same.

```
magick -size 352x80 -depth 8 raw.yuv x.png
```
x.png is:
![x](https://user-images.githubusercontent.com/33812368/234304163-9d8db8ae-8f76-48c3-93c2-df093651ab53.png)

Is that the image you expect?

The defaults are interlace = 2 (plane),  horizontal_factor=2, and vertical_factor=2.

We can convert x.png back to YUV:
```
magick x.png raw2.yuv
```
The files raw.yuv and raw2.yuv are the same size, 42240 bytes, and corresponding bytes are mostly identical.

... then compare the images in the two YUV files:
```
magick -size 352x80 -depth 8 raw.yuv raw2.yuv -metric RMSE -format %[distortion] -compare info:

0.019571998
```
On a scale of 0.0 to 1.0, where 0.0 means "identical", the RMSE is 0.01957. As the file is 8-bit, with two channels downsampled by 2x2 (ie lossy compression, 4 pixels are compressed into 6 bytes), we can't expect better.

So the round-trip seems to be a success. The output raw2.yuv is virtually the same as the input raw.yuv.

**Comment by Eutechnyx** [comment]
Nope. Obviously the image i expect from that file looks like this:
[Correct image](https://drive.google.com/file/d/1GdvZNmX08Yg-GnY2q7qQmagWOR8CBLbN/view?usp=share_link)
Image you got was converted as from planar YUV420.

From what i understood, it is a rare format of YUV420, where Y component is planar, and UV's are interleaved.
Magick is converting it only as a planar component values, no matter if `-interlace plane` option was set or not.

Found that it can be correctly decoded using ffmpeg, with  pixel_format set to nv12:
`ffmpeg -pixel_format nv12 -video_size 352x80 -i raw.yuv output.png`

Does magick support such pixel output method?

**Comment by snibgo** [comment, marked as answer]
> From what i understood, it is a rare format of YUV420, where Y component is planar, and UV's are interleaved.

IM does not contain a simple option for that.

Perhaps a complex series of IM operations could be used to read and write that format.

Or add an entry in delegates.xml to call ffmpeg to do the conversion.

**Reply by Eutechnyx** [reply]
Thanks, made that convertion as a a part of python script:

y_size = width * height
u_size = v_size = y_size // 4
y_plane = yuv420_data[:y_size]
u_plane = yuv420_data[y_size:y_size+u_size]
v_plane = yuv420_data[y_size+u_size:y_size+u_size+v_size]

uv_plane = bytearray(u_size+v_size)
for i in range(u_size):
    uv_plane[2*i] = u_plane[i]
    uv_plane[2*i+1] = v_plane[i]

nv_data = y_plane + uv_plane
